### ISSUE TITLE
[8.0] Missing journal entries when Dropship is used with real time automated Inventory method

### ISSUE THREAD

**Issue by sdalexander** [odoo/odoo#12687]
Impacted versions: 
8.0

Steps to reproduce:
Set up Product as stockable
set inventory valuation to real time automated
set all account stock properties on the Product category.
Enter a sales order using the above product and choose route of Dropship and set invoice policy to 'on demand.
Confirm sales order , create sales invoice ,confirm purchase order , receive goods then generate supplier invoice.

Current behavior:
As the sale & purchase Invoice journal transactions happen as normal, using the Stock Input account and the Stock output account from the Product/Product category.
But the delivery is from ‘Partners/suppliers’ to ‘Partners/customers’, it does not hit stock and make any accounting entries.
 This leaves the ‘210050 GRNI’ with a debit entry outstanding and the ‘100120 GDNI’ with a Credit Entry outstanding.

Expected behavior:
To correct this we need a Dropship Stock Journal that will debit the Stock Input account and Credit the Stock Output account against the product.
This will balance out the outstanding entries, When the goods are shipped.

see attached for more detail
[dropship COS test.pdf](https://github.com/odoo/odoo/files/347622/dropship.COS.test.pdf)

[Dropship accounting issue.pdf](https://github.com/odoo/odoo/files/347621/Dropship.accounting.issue.pdf)

Video/Screenshot link (optional):

**Comment by Yenthe666** [comment]
@qdp-odoo looks like one for you!

**Comment by sdalexander** [comment]
Hi guys

I wonder if you could have a look at whether you think this is a bug or whether there is a workaround that I am missing?

regards

Steve

**Comment by max3903** [comment]
@sdalexander I just did your configuration and went through the whole dropshipping and normal processes. 
* Dropship: Nothing hit my stock journal and stock accounts
* Normal: 2 accounting entries (in and out).

**Comment by sdalexander** [comment]
Certainly a dumb comment.have you tested it?  On 26 July 2016 at 07:12 Ermin Trevisan <notifications@github.com> wrote:Maybe I'm dumb, but the last 25 years I thought that the essence of the term "Dropshipping" is that it does not touch your stock.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by sdalexander** [comment]
Hi @max3903 

yes that is correct, 
I did not install the anglo-accounting manually , I think it comes in with one of the setup options. 
I thought it was part of the 'real time stock valuation set up?

can you have real time stock valuation without anglo-saxon accounting?

**Comment by treviser** [comment]
@sdalexander That's why I deleted it immediately after posting it erraneously.

**Comment by sdalexander** [comment]
Hi ErminOk , I will do the same, I agree with you that Dropship should not touch the stock accounts.but if you configure for real time inventory on an Item that you usually hold in stock, then on a particular order, choose Dropshipping on the order line for that product. When you create the invoice it will  create accounting entries from the stock output account you have specified for that item. But as you have dropshipped the item, as you rightly point out no stock entries are done, leaving you with an imbalance on the stock output account. As I don't think you can stop the Invoice entry happening then I am suggesting that if dropship is used on an item with inventory valuation of 'real time (automated)' it should create the balancing entries for the stock output account.regardsSteveOn 26 July 2016 at 09:21 Ermin Trevisan <notifications@github.com> wrote:@sdalexander That's why I deleted it immediately after posting it erraneously.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by max3903** [comment]
@sdalexander 
Real-time inventory allows you to have your stock evaluation in the accounting synchronized with your WMS and your stock level. Otherwise your stock value in your accounting reports is "wrong" until you count and create an adjustment entry.

Anglo-saxon accounting:
The difference between the Anglo-Saxon accounting countries and the Rhine (or also called Continental accounting) countries is the moment of taking the Cost of Goods Sold versus Cost of Sales. Anglo-Saxons accounting does take the cost when sales invoice is created, Continental accounting will take the cost at the moment the goods are shipped.
https://apps.openerp.com/apps/7.0/account_anglo_saxon/

**Comment by sdalexander** [comment]
@max3903 

thanks for that, In the UK, cost of sales is usually taken at Invoice creation , so that the sale and the cost are then always together in the same period.
do you think the anglo_saxon module can be fixed so if dropship is used on an item with inventory valuation of 'real time (automated)' it should create the balancing journal and  entries for the stock output account.?

**Comment by sdalexander** [comment]
@Yenthe666 @max3903 

Hi Guys

So what do you think is the best way to go about getting this fixed,  Is it a bug with in the Anglo saxon side ?

regards
Steve

**Comment by max3903** [comment]
@sdalexander @Yenthe666 I have asked my customer to report it to Odoo SA with a link to this issue using his Enterprise contract. I suggest you to do the same or propose a fix.

**Comment by sdalexander** [comment]
Thanks @max3903 
I will report it, and yes I am producing my customer a  sql query to identify the transactions for him to journal manually.
Not a great solution in there eyes.

**Comment by qdp-odoo** [comment]
Hi there,

for your information, the bug has been confirmed and the patch is soon to be implemented. I suppose that the OPW creator will report here as well?

**Comment by max3903** [comment]
@qdp-odoo The customer is going live in the next few weeks. Not sure they want to take the risk of a last minute change and repeat tests.

**Comment by bistaray** [comment]
Can you share the fix (I have an OPW customer and we are not getting correct entries for Drop Ship when using Anglo Saxon).

**Comment by spadae22** [comment]
@qdp-odoo We are on 8 also and have an enterprise contract.

**Comment by sdalexander** [comment]
That's great news @qdp-odoo. how do we get the patch?

**Comment by chuckdeblois** [comment]
following

**Comment by bistaray** [comment]
Update: I have been in touch with Odoo support.  They tell me the patch has not yet been finished and have added me as a follower to the odoo.com OPW Issue 684,742 - "Bug in dropship journal moves"

**Comment by sdalexander** [comment]
thanks for the update @bistaray

**Comment by bistaray** [comment]
Cleaning up transactions now.  I created a new liability account "211200 Drop Shipped Not Vouchered".

For Drop Ship Vendor Invoices:

`select number from account_invoice where origin in (select name from purchase_order where dest_address_id is not null and state not in ('draft','sent'));`

I am changing the invoice lines and debit journal items that are currently booking to "211200 Received Not Vouchered" and sending them to "Drop Shipped Not Vouchered":

`update account_invoice_line set account_id = (select id from account_account where code = '211201' and company_id = 1) where account_id = (select id from account_account where code = '211200' and company_id = 1) and invoice_id in (select id from account_invoice where origin in (select name from purchase_order where dest_address_id is not null and state not in ('draft','sent')));`

and

`update account_move_line set account_id = (select id from account_account where code = '211201' and company_id = 1) where account_id = (select id from account_account where code = '211200' and company_id = 1) and move_id in (select move_id from account_invoice where id in (select id from account_invoice where origin in (select name from purchase_order where dest_address_id is not null and state not in ('draft','sent'))))`

For Drop Ship Customer Invoices:

I am changing the credit entries that are currently booking to "121025 Uninvoiced Delivered Goods / Delivered Not Vouchered" and sending them to "211201 Drop Shipped Not Vouchered".

`update account_move_line set account_id = (select id from account_account where code = '211201') where account_id = (select id from account_account where code = '121025') and move_id in (select move_id from account_invoice where id in (select id from account_invoice where route = 47 and state <> 'cancel'))`

(47 is the ID of the DROP SHIP route from the stock_location_route table - note that route is a custom field in the database I am doing this work - populated if the SO is a drop ship.  Without this I would probably look for Invoices from SO's where SO lines where route = 47.  I am lucky that this Customer doesn't mix dropshipping and standard fulfilment.)

**Comment by qdp-odoo** [comment]
god damn.... 

just reverted my fix on 9.0 because I realize it was
- needed for 8.0
- badly placed because purchase doesn't depend on sale (anymore)

**Comment by qdp-odoo** [comment]
The fix has just landed in v8 in revision https://github.com/odoo/odoo/commit/7bdd4de8051afa11e3941f7ab930c96b2e82bbc7 and will soon be forward-ported to next versions.

As you'll notice, there's a new module (anglo_saxon_dropshipping) to install in order to fix that behavior. This is only for technical reasons (modules dependancies to stock_dropshipping and account_anglo_saxon) and has no other side effect than fixing this bug. It will also install automatically on new instances.

The solution implemented is to check if the customer invoice line is related to sale order lines having one of its procurement_ids with a purchase_line_id set. If yes, it means that it is a confirmed dropship and in that case we don't call to super (we don't create the cost of sale line). The drawback is that it won't work if the procurement is in exception at the customer invoice time,(the behavior will be as it is currently), but I consider this ok since we don't know how the procurement will be solved (it may not be a dropshipping), and it'll be only at the beginning (once the config is done it shouldn't go in exception anymore). People will have to manually fix those amounts with a miscellaneous operation.

This patch also does not do anything for the supplier invoice increasing the GRNI (stock input account) : indeed, users in anglo saxon mode should not use the 'stock interim account (received)' on supplier invoices for dropshipped goods, but rather use the COGS directly on the invoice line.

Thanks for the contribution

**Comment by sdalexander** [comment]
Hi  @qdp-odoo 
I am not sure that this fully fixes the problem that was reported . As one of the main reasons for using 'Anglo saxon'  and 'real time automated' is to ensure that both the sale and cost of sale are recorded at the same time, 
as well as the keeping a stock value for items that are usually kept in stock but  can be overridden at the sale order line level.
So to stop the COS line being created breaks this.
This now means we have a sale without an associated COS ... and potentially users will have to manually change supplier invoices from 'GRNI' to the Expense account which is not really feasible.

I hope you will reconsider your solution

**Comment by qdp-odoo** [comment]
@sdalexander Excuse me by advance if I sound harsh, it's not meant at all and be sure that I only want to find the best solution for this problem.... but I'm gonna reply as the arguments are coming to my mind ^^


So although I understand you, I'd like some other point of view here, because I'm really not convinced. 

1) yes it's True that the sales and Cost Of Sales aren't recorded in the same time for those operations, meaning you have a bit of latency... but
 * that's more a matter of reporting... rather than saying it is a functional bug, some may have to adapt a little bit their way to work
 * that's how you work with subcontracting, don't you? Sucontracting a a kind of dropshipped service, but you register the COS at the time of the PO, not the SO, hu?
 * that's actually how continental accounting works for _all_ operations, so I believe that either it isn't hard to get those figures, either this latency is negligible

2) "potentially users will have to manually change supplier invoices from 'GRNI' to the Expense account which is not really feasible".
 * Its a default value, it's meant to be manually changed...
 * Although I agree it's not convenient, it looks to me it's more like a usability problem than a functional bug... And we may fix that default value in a future version

3) Actually, the other solution was to create stock valuation entry for stock move of dropshipped goods:
 * That doesn't sound logical at all to me: maybe I'm wrong, but I don't see how a stock move that doesn't physically transit through your stock would create a stock valuation accounting entry.
 * What about the use case where a customer of you request a replacement piece that you don't have, you would just create a PO with the destination being the customer... Here, creating the valuation entry for the stock move would be wrong....

4) A possible workaround is to create an internal location for dropship and configure pull/push rules to forward all goods requested/received to supplier/customer. That would have to behavior you'd expect but would pollute the accounting entries, wouldn't it?

I just had an internal discussion about that and my colleague's opinion was even to not change anything at all: accountants would have to manually create accounting entries that zero out the GRNI and GDNI periodically... So it seems we definitively need more opinions here!

**Comment by max3903** [comment]
I think we have 2 solutions:

1. Nothing hit the input and output accounts for dropshipped item when invoices are validated. Like in the continental accounting.
2. The dropship stock move generates an accounting entry to zero input and output accounts.

**Comment by sdalexander** [comment]
Hi @qdp-odoo 

thanks for your reply, I am happy to have an open discussion on this and to get more input... 

My issue is mainly that it is not a good story for customers wanting anglo saxon accounting but having
certain transactions that will default back to continental accounting and then having to give them a manual workaround.

It makes accounting and auditing difficult in the UK when these transaction fall across period and year end  boundaries. 
On the PO side it is not always easy to tell which invoices relate to a dropship for a product that is set up for stock as well.

I did try your suggestion point 4 push & pull rules ... but I couldn't get that to work in practice 
to post to the correct accounts. ( If anybody has that working let me know)

I do like the 2nd suggestion from @max3903 
'The dropship stock move generates an accounting entry to zero input and output accounts.'
better than asking people to do that manually.

**Comment by qdp-odoo** [comment]
and what if the default account on supplier invoice was set to the expense account in case of dropshipping?

**Comment by max3903** [comment]
@qdp-odoo input stock account can't be change on the invoice.

**Comment by sdalexander** [comment]
Hi @qdp-odoo 
what about  @max3903 's  point that we can't change that input stock account on the PO?

I still feel that if anglo saxon is installed then the cost of sale should be taken at the invoice point.

could we have a separate dropship account so that it doesnt pollute the Stock one?

**Comment by bistaray** [comment]
>>@qdp-odoo input stock account can't be change on the invoice.

I agree - this is not exposed for the user, it comes from the Product Category.

------------------------------------------------------------------------------------------

I have discussed the overall issue with the Customer it is affecting, a personal Accounting (CFO) colleague and a Bista colleague.

We'd *LIKE* for the COGS to be taken at the point of the Customer Invoice, without any manual intervention.  We all question the accounting implications of Invoicing a Customer without COGS being booked, not to mention margin calculations on that Invoice (since the Sales Order happens before the true cost of the product is known in a real-price scenario).

We'd *LIKE* for a 'DROPSHIP - PRODUCT VALUE/COGS - INPUT AND OUTPUT ACCOUNT**' (on Product Category) to be used instead of the stock input and output accounts.  This would allow for a doublecheck to see if the value invoiced from the Vendor matches the COGS cost invoiced to the Customer.  It should, of course, but with real pricing it likely won't if the product is also purchased and received in the warehouse.

** Dropship Stock Account - is what we used when cleaning up the already processed entries for Drop Ship.

The "Uninvoiced Received Goods" entry necessary at the time of posting the Vendor Invoice (to offset the CREDIT to ACCOUNTS PAYABLE) represents "PRODUCT VALUE not yet Invoiced to the Customer" or "COGS yet to be posted" or "uninvoiced COGS" and would work fine if it was posted to "DROPSHIP STOCK ACCOUNT".

The "Uninvoiced Delivered Goods" entry necessary at the time of posting the Customer Invoice (to offset the DEBIT to ACCOUNTS RECEIVABLE) represents "PRODUCT VALUE now Invoiced to the Customer" or "PRODUCT VALUE transferred to COGS" and would work find if it was posted from "DROPSHIP STOCK ACCOUNT"

My Customer intends to post the Vendor Invoice first, then immediately after this, the Customer Invoice.  Their feedback on manually changing anything was that it was error prone.

Please keep us updated on what you plan or don't plan to do, so we can take over if needed.

Appreciate the work already done and the open dialog!

**Comment by sdalexander** [comment]
Thanks @bistaray  for the detailed response I think this is would be solution to the problem.

**Comment by sdalexander** [comment]
Hi @bistaray @qdp-odoo @max3903 
Has there been any progress on this issue? or do we just have to get it fixed ourselves.
as per @bistaray 's  solution above

regards
Steve

**Comment by bistaray** [comment]
@sdalexander I asked for an update on the 19th and it was explained to me that this is a complex issue and @qdp-odoo is in discussions with other experts for a resolution to how to close this (which is why it is still open).  I asked for another update today.

**Comment by sdalexander** [comment]
Hi Ray

Thanks for the update,

Regards 

Steve


Director
Opensys Ltd
07964 282551

**Comment by bistaray** [comment]
Update:  Bista has Odoo Support Ticket#690561 to track this.

They gave us an update today:

_The issue had been fixed in 8.0 with this commit : https://github.com/odoo/odoo/commit/7bdd4de8051afa11e3941f7ab930c96b2e82bbc7
I've discussed with Quentin and he has learned afterwards that apparently this fix was not the correct way of doing it, that's why it's not in 9.0 right now.
He will soon commit a better fix for both 8.0 and 9.0.
You will be informed as soon as it's done, I keep this ticket opened._

**Comment by max3903** [comment]
I discussed the issue with @qdp-odoo and @fhe-odoo during Odoo Experience last week. We are waiting for @fhe-odoo to contact Odoo's accountant to validate the solution.

**Comment by qdp-odoo** [comment]
Here's the new version of the fix, soon to be deployed on 8.0 I suppose (waiting the redbot to be happy)... I didn't had the confirmation of fhe's accountant but as you all agree saying my previous fix was wrong, I surrendered ^^

You can test it on runbot (branch 8.0-staging-qdp) in the meantime.
Thanks for your contribution and (especially for this one) for your patience!

**Comment by max3903** [comment]
:+1: LGTM

**Comment by sdalexander** [comment]
@qdp-odoo  Thanks for that, it is much appreciated. especially by the customer...

I will test it and let you know.

also thanks to @max3903  and @bistaray

**Comment by qdp-odoo** [comment]
Finally, the bugfix landed on 8.0 at revision 001610e

Let's hope this close that topic for good! :-)

**Comment by bistaray** [comment]
Is a v9 patch coming also?

**Comment by qdp-odoo** [comment]
it just has been forward-ported to 9.0 at revision https://github.com/odoo/odoo/commit/b14c8b361ce8223b7dc277797dd501e00907c19f

**Comment by bistaray** [comment]
💯

**Comment by Yenthe666** [comment]
Great job @qdp-odoo! 👍 for you.

**Comment by alan196** [comment]
@qdp-odoo I just try your fix and it makes the account entry the dropshipping stock move 👍 . The problem now is in the sales invoice. Your previous fix https://github.com/odoo/odoo/commit/7bdd4de8051afa11e3941f7ab930c96b2e82bbc7 is now making the same but now in Stock Output Account.

**Comment by qdp-odoo** [comment]
@alan196 don't consider the fix 7bdd4de but only 001610ece9cf20e8.

**Comment by bistaray** [comment]
@qdp-odoo - I just noticed that when I visit https://github.com/odoo/odoo/commit/b14c8b361ce8223b7dc277797dd501e00907c19f I see v10.  Is that the right link?

![](http://i.imgur.com/03WMPVE.png)

**Comment by qdp-odoo** [comment]
yes it is, don't pay attention to the branch noticed by github. I suppose it means it has been forward-ported up to the branch 10.0.

Patch for v8.0: 001610ece9cf20e
Patch for v9.0 and further: b14c8b361ce (only the part in addons/stock_account/stock_account.py)

**Comment by alan196** [comment]
@qdp-odoo sorry the commit https://github.com/odoo/odoo/commit/001610ece9cf20e8f6ff3e928f18ed0f7fe6c9ff is for odoo 8.0, I just tested with the latest version of Odoo 9.0 and when you create the customer invoice the account is not correct.

**Comment by qdp-odoo** [comment]
sorry, what account are you talking about? what is the use case not working according to you?

**Comment by alan196** [comment]
In Odoo 9.0 
- When you confirm a purchase order the dropshipping picking is created, and when you validate it. It will create an journal entry with the **stock input account** and the **stock output account**. (This is **correct**).
- When you create supplier invoice the **stock input account** is charged. (This is **correct**).
- When you create customer invoice the cost **stock output account** is not in the journal entry. (This is **incorrect**). For that reason the **stock output account** is unbalanced.

**Comment by qdp-odoo** [comment]
geez... got it :palmface:

5a77ff6 fix it for good in v9... I forgot to remove the code from my previous attempt when I forward-ported my second fix.

Sorry for the inconvenience. @alan196 can you confirm it's working this time?

**Comment by gdgellatly** [comment]
@qdp-odoo I'm not saying whether I agree or disagree with this fix.  In
truth, I think both sides are arguable, and many years ago I argued for
this very fix, but eventually came to accept that it wouldn't happen, and
so rewrote anglo saxon myself.  But please, given the amount of change and
debate this module has had we need tests for every scenario and we need
good documentation.  Anyone implementing this will have different entries
as a result, definitely changing behaviour of a stable version.

I just realised I have not pushed code, however it is irrelevant for now,
where I wrote tests for most scenarios in anglosaxon.  I used hypothesis as
the basis for this test, and even things I thought worked, definitely
failed in edge cases.  Most of the failures were related to the stock
module, rather than account or anglo_saxon directly and usually involved
either the zero case or rounding.  That was v7 and I will need to do same
exercise for 10, but without documented expected behaviour, or any tests
for such a critical module other than very simple cases, where does one
start?
